gnark-prover/prover/circuit_utils.go

```diff
@@ -63,6 +78,25 @@ func (gadget InclusionProof) DefineGadget(api frontend.API) interface{} {
 	return currentHash
 }
 
+func (gadget NonInclusionProof) DefineGadget(api frontend.API) interface{} {
+	currentHash := make([]frontend.Variable, gadget.NumberOfUtxos)
+	for proofIndex := 0; proofIndex < gadget.NumberOfUtxos; proofIndex++ {
+		api.AssertIsDifferent(gadget.LeafLowerRangeValue[proofIndex], gadget.Value[proofIndex])
+		api.AssertIsLessOrEqual(gadget.LeafLowerRangeValue[proofIndex], gadget.Value[proofIndex])
+
+		api.AssertIsDifferent(gadget.LeafHigherRangeValue[proofIndex], gadget.Value[proofIndex])
+		api.AssertIsLessOrEqual(gadget.Value[proofIndex], gadget.LeafHigherRangeValue[proofIndex])
+
+		currentHash[proofIndex] = abstractor.Call(api, poseidon.Poseidon3{In1: gadget.LeafLowerRangeValue[proofIndex], In2: gadget.LeafIndex[proofIndex], In3: gadget.LeafHigherRangeValue[proofIndex]})
+
+		currentPath := api.ToBinary(gadget.InPathIndices[proofIndex], gadget.Depth)
```

Comment:
why don't we use a verify Merkle proof sub circuit so that we don't have duplicate code in the inclusion and non-inclusion proofs?

Reply:
Refactored it https://github.com/Lightprotocol/light-protocol/pull/559/files#diff-d0aaaedb6d35ac5ecee522b2451bce7e12976bfaddbe4ee3cbe257bd2b9ad19aR102-R130